File: langs/csharp-ddd/src/MartianDDD.Domain/Robot.cs
namespace MartianDDD.Domain;

public sealed class Robot
{
    public Position Position { get; private set; }
    public Orientation Orientation { get; private set; }
    public bool Lost { get; private set; }

    public Robot(Position position, Orientation orientation)
    {
        Position = position;
        Orientation = orientation;
        Lost = false;
    }

    public void TurnLeft() => Orientation = Orientation switch
    {
        Orientation.N => Orientation.W,
        Orientation.W => Orientation.S,
        Orientation.S => Orientation.E,
        Orientation.E => Orientation.N,
        _ => Orientation
    };

    public void TurnRight() => Orientation = Orientation switch
    {
        Orientation.N => Orientation.E,
        Orientation.E => Orientation.S,
        Orientation.S => Orientation.W,
        Orientation.W => Orientation.N,
        _ => Orientation
    };

    public Position NextForward() => Orientation switch
    {
        Orientation.N => new Position(Position.X, Position.Y + 1),
        Orientation.E => new Position(Position.X + 1, Position.Y),
        Orientation.S => new Position(Position.X, Position.Y - 1),
        Orientation.W => new Position(Position.X - 1, Position.Y),
        _ => Position
    };

    public void MoveTo(Position p) => Position = p;
    public void MarkLost() => Lost = true;
}

File: python-martian-ddd/src/martian_ddd/domain/robot.py
from __future__ import annotations
from dataclasses import dataclass
from .types import Position, Orientation


@dataclass
class Robot:
    position: Position
    orientation: Orientation
    lost: bool = False

    def turn_left(self) -> None:
        self.orientation = {
            Orientation.N: Orientation.W,
            Orientation.W: Orientation.S,
            Orientation.S: Orientation.E,
            Orientation.E: Orientation.N,
        }[self.orientation]

    def turn_right(self) -> None:
        self.orientation = {
            Orientation.N: Orientation.E,
            Orientation.E: Orientation.S,
            Orientation.S: Orientation.W,
            Orientation.W: Orientation.N,
        }[self.orientation]

    def next_forward(self) -> Position:
        match self.orientation:
            case Orientation.N:
                return Position(self.position.x, self.position.y + 1)
            case Orientation.E:
                return Position(self.position.x + 1, self.position.y)
            case Orientation.S:
                return Position(self.position.x, self.position.y - 1)
            case Orientation.W:
                return Position(self.position.x - 1, self.position.y)
        return self.position

